File: ios/Talise/Features/Home/HistoryRow.swift
import SwiftUI

/// Single history row. Reused by Home (top 4) and HistoryView (full list).
///
/// Visual treatment is the Liquid Glass card from TaliseGlassCard with
/// a directional tint stacked over the system material:
///   • Sent     → small red tint
///   • Received → small green tint
///   • Other    → no tint (neutral glass)
///
/// Tints are intentionally low alpha (~0.10) so the row reads as
/// "subtly colored glass" against the dark page, not a solid pill.
struct HistoryRow: View {
    let entry: ActivityEntryDTO
    let onTap: () -> Void
    /// Optional callback fired when the user taps the "Swap to USDsui"
    /// CTA that appears on inbound non-USDsui coin rows (2026-05-29 —
    /// replaces the archived auto-swap cron). When nil, the CTA is
    /// hidden and the row behaves as before.
    var onSwapToUsdsui: (() -> Void)? = nil

    /// True when the inbound coin on this row should surface the
    /// "Swap to USDsui" affordance. Triggers on:
    ///   • direction == "received"
    ///   • otherCoin present (i.e. NOT a plain USDsui/SUI receive)
    ///   • otherCoin.symbol is not already USDsui
    private var showsSwapCTA: Bool {
        guard onSwapToUsdsui != nil else { return false }
        guard entry.direction == "received" else { return false }
        guard let other = entry.otherCoin else { return false }
        return other.symbol.uppercased() != "USDSUI"
    }

    var body: some View {
        Button(action: onTap) {
            HStack(spacing: 14) {
                ZStack {
                    // Tinted directional badge — dusty red for Sent,
                    // mossy green for Received, accent for Invest. The
                    // bg is the tint at ~32% over the page bg, the
                    // arrow is the tint at full saturation so it
                    // reads as a colored glyph on a colored disc.
                    Circle()
                        .fill(badgeBgColor)
                        .frame(width: 32, height: 32)
                    Image(systemName: iconName)
                        .font(.system(size: 13, weight: .medium))
                        .foregroundStyle(badgeFgColor)
                }
                VStack(alignment: .leading, spacing: 2) {
                    Text(title)
                        .font(TaliseFont.body(13, weight: .light))
                        .kerning(-0.48)
                        .foregroundStyle(TaliseColor.fg)
                    MicroLabel(text: subtitle, color: TaliseColor.fgDim)
                        .kerning(-0.32)
                    if showsSwapCTA {
                        // "Swap to USDsui" — small, accent-tinted CTA
                        // shown directly under the subtitle on inbound
                        // non-USDsui coin rows. Replaces the archived
                        // auto-swap cron. POSTs `/api/swap/prepare` via
                        // the caller-supplied `onSwapToUsdsui` handler.
                        Button {
                            onSwapToUsdsui?()
                        } label: {
                            HStack(spacing: 4) {
                                Image(systemName: "arrow.triangle.2.circlepath")
                                    .font(.system(size: 10, weight: .semibold))
                                Text("Swap to USDsui")
                                    .font(TaliseFont.body(11, weight: .medium))
                            }
                            .foregroundStyle(TaliseColor.accent)
                            .padding(.top, 2)
                        }
                        .buttonStyle(.plain)
                    }
                }
                Spacer()
                // Amount only — the whole row is tappable so the
                // "Details ↗" eyebrow was visual filler. Subtitle
                // already carries the "tap me" affordance via the
                // row-press tint.
                Text(amountFormatted)
                    .font(TaliseFont.body(14, weight: .light))
                    .kerning(-0.56)
                    .foregroundStyle(TaliseColor.fg)
            }
            .padding(.horizontal, 16)
            .padding(.vertical, 12)
            .contentShape(Rectangle())
        }
        // Directional tint is only applied while the user is pressing
        // the row — at rest the row reads as neutral glass; on press
        // it picks up red (sent) or green (received) with a smooth
        // crossfade. Neutral category never tints.
        .buttonStyle(HistoryRowButtonStyle(
            tintColor: tintColor,
            tintAlpha: tintAlpha
        ))
    }

    // MARK: - Category + tint

    private enum Category {
        case sent
        case received
        case invest
        case withdraw
        case autoswap
        case neutral
    }

    /// Server-side `direction` field carries the classification. Plain
    /// transfers ride the chain-derived sent/received; yield venue
    /// txs (DeepBook supply, NAVI lending) come back as
    /// `invest`/`withdraw`; vault auto-swap conversions come back as
    /// `autoswap` (emitted by `VaultAutoSwap` event in the Move
    /// `talise::vault` module). Each gets its own icon + tint.
    private var category: Category {
        switch entry.direction {
        case "received": return .received
        case "invest":   return .invest
        case "withdraw": return .withdraw
        case "autoswap": return .autoswap
        // DEX swaps (the legacy Convert banner, manual Cetus calls,
        // anything where the user moved one coin in and a different
        // one out in the same tx) share the auto-swap visual
        // language — green leaf, accent tint — because the user
        // doesn't care which path moved the funds; what matters is
        // "I converted X into Y." Auto-swap and manual swap render
        // identically.
        case "swap":     return .autoswap
        case "sent":     return .sent
        default:         return .neutral
        }
    }

    private var tintColor: Color {
        switch category {
        case .sent:     return Color(hex: 0xC95A4A)
        case .received: return Color(hex: 0x4FB35E)
        // Invest + withdraw share the Talise green accent — they're
        // yield motion, not money lost / gained from another wallet.
        case .invest:   return TaliseColor.accent
        case .withdraw: return Color(hex: 0x4FB35E)
        // Auto-swap also reads as green — it's the system working on
        // the user's behalf to keep the @handle in USDsui.
        case .autoswap: return TaliseColor.accent
        case .neutral:  return .clear
        }
    }

    /// Circular badge fill (the disc behind the arrow). A 32% wash of
    /// the directional color over the page background reads as
    /// "muted dusty red / forest green" without competing with the
    /// row's content text.
    private var badgeBgColor: Color {
        switch category {
        case .sent:     return Color(hex: 0xC95A4A).opacity(0.32)
        case .received: return Color(hex: 0x4FB35E).opacity(0.32)
        case .invest:   return TaliseColor.accent.opacity(0.22)
        case .withdraw: return Color(hex: 0x4FB35E).opacity(0.32)
        case .autoswap: return TaliseColor.accent.opacity(0.22)
        case .neutral:  return TaliseColor.surface2
        }
    }

    /// Arrow color inside the badge. Sits at full saturation against
    /// the muted disc, matching the design ref where the glyph is
    /// noticeably brighter than its background.
    private var badgeFgColor: Color {
        switch category {
        case .sent:     return Color(hex: 0xF0A99E)
        case .received: return Color(hex: 0xA9DFB3)
        case .invest:   return TaliseColor.accent
        case .withdraw: return Color(hex: 0xA9DFB3)
        case .autoswap: return TaliseColor.accent
        case .neutral:  return TaliseColor.fg
        }
    }

    private var tintAlpha: Double {
        switch category {
        case .sent, .received, .invest, .withdraw, .autoswap: return 0.18
        case .neutral:                                         return 0
        }
    }

    /// SF Symbol used in the circular badge. Invest uses the leaf
    /// (matches the Invest tab bar icon, so the connection between
    /// "the tab I supplied from" and "this row" is visual not just
    /// textual). Withdraw mirrors with the leaf inverted via
    /// arrow.down-on-leaf — see invest case below.
    private var iconName: String {
        switch category {
        case .sent:     return "arrow.up.right"
        case .received: return "arrow.down.left"
        case .invest:   return "leaf.fill"
        case .withdraw: return "leaf"
        // Auto-swap reuses the leaf — same family as the Earn / Invest
        // tab, signalling "the system worked for you". The conversion
        // is implicit in the title ("Auto-swapped 0.5 SUI → $1.20").
        case .autoswap: return "leaf.fill"
        case .neutral:  return "circle"
        }
    }

    private var title: String {
        // Non-USDsui/non-SUI rows (WAL, USDC, USDT, …) override the
        // default "Sent"/"Received" so the row clearly shows the coin.
        if let other = entry.otherCoin {
            return entry.isReceived
                ? "Received \(other.symbol)"
                : "Sent \(other.symbol)"
        }
        switch category {
        case .sent:     return "Sent"
        case .received: return "Received"
        case .invest:
            if let v = entry.venue, !v.isEmpty {
                return "Invested in \(displayVenueName(v))"
            }
            return "Invested"
        case .withdraw:
            if let v = entry.venue, !v.isEmpty {
                return "Withdrew from \(displayVenueName(v))"
            }
            return "Withdrew"
        case .autoswap:
            // Two flavors share this case:
            //   • direction == "autoswap" → the vault's cron-driven
            //     swap (sub-second auto-conversion). Server emits the
            //     source coin via `venue`.
            //   • direction == "swap" → any DEX swap touching the
            //     user's wallet (legacy Convert banner, direct Cetus
            //     call, etc.). For these we render "Swapped X → Y"
            //     using the SUI / otherCoin / USDsui legs we have.
            //
            // Title is just the verb; `amountFormatted` does the
            // "X → Y" composition.
            if entry.direction == "swap" { return "Swapped" }
            if let v = entry.venue, !v.isEmpty {
                return "Auto-swapped \(v.uppercased())"
            }
            return "Auto-swapped to USDsui"
        case .neutral:  return "Activity"
        }
    }

    private var subtitle: String {
        let date = Date(timeIntervalSince1970: entry.timestampMs / 1000)
        let fmt = RelativeDateTimeFormatter()
        fmt.unitsStyle = .abbreviated
        return fmt.localizedString(for: date, relativeTo: Date())
    }

    private var amountFormatted: String {
        // Auto-swap & manual swap are net-neutral economically — one
        // coin in, a different coin out. We render BOTH legs of the
        // transformation ("0.1 SUI → ₦139.59") so the row reads as
        // a conversion rather than a debit/credit. The title already
        // tells the user what category they're looking at.
        if category == .autoswap {
            // Build the leg strings independently so we can compose
            // "from → to" no matter which fields the server populated:
            //   • SUI ↔ USDsui swap: amountSui + amountUsdsui
            //   • USDC/WAL/etc → USDsui: otherCoin + amountUsdsui
            //   • USDsui → SUI (rare): amountUsdsui + amountSui
            var legs: [String] = []
            if let sui = entry.amountSui, sui > 0 {
                legs.append(String(format: "%.4f SUI", sui))
            }
            if let other = entry.otherCoin {
                legs.append("\(other.displayAmount) \(other.symbol)")
            }
            if let usd = entry.amountUsdsui, usd > 0 {
                legs.append(TaliseFormat.local2(usd))
            }
            switch legs.count {
            case 0: return "→ —"
            case 1: return "→ \(legs[0])"
            default:
                // Always end on USDsui when present — that's the
                // canonical Talise unit, and the "destination" the
                // user opted into when they enabled auto-swap.
                return "\(legs[0]) → \(legs[1])"
            }
        }
        // Invest = wallet → pool (debit, "-"); Withdraw = pool → wallet
        // (credit, "+"). Plain transfers use direction directly.
        let isInflow = entry.isReceived || entry.isWithdraw
        let prefix = isInflow ? "+" : "-"
        // Non-USDsui/non-SUI row: format raw u64 with the coin's
        // decimals + symbol. We don't compute a USD value because
        // we don't have a reliable price for arbitrary tokens.
        if let other = entry.otherCoin {
            return "\(prefix)\(other.displayAmount) \(other.symbol)"
        }
        if let usd = entry.amountUsdsui {
            return prefix + TaliseFormat.local2(Swift.abs(usd))
        }
        if let sui = entry.amountSui {
            return String(format: "\(prefix)%.4f SUI", Swift.abs(sui))
        }
        return prefix + "—"
    }
}

/// Glassy row background that animates a directional tint in/out
/// based on the button's pressed state. At rest the row reads as
/// neutral glass; on press it picks up red (Sent) or green (Received).
private struct HistoryRowButtonStyle: ButtonStyle {
    let tintColor: Color
    let tintAlpha: Double

    func makeBody(configuration: Configuration) -> some View {
        let shape = RoundedRectangle(cornerRadius: 18, style: .continuous)
        let alpha = configuration.isPressed ? tintAlpha : 0
        return configuration.label
            .background(
                ZStack {
                    shape.fill(.ultraThinMaterial)
                    shape.fill(Color.black.opacity(0.35))
                    shape.fill(tintColor.opacity(alpha))
                }
            )
            .overlay(
                shape.strokeBorder(
                    LinearGradient(
                        colors: [
                            Color.white.opacity(0.16),
                            Color.white.opacity(0.04),
                            Color.white.opacity(0.08),
                        ],
                        startPoint: .top,
                        endPoint: .bottom
                    ),
                    lineWidth: 1
                )
            )
            .clipShape(shape)
            .scaleEffect(configuration.isPressed ? 0.985 : 1.0)
            .shadow(color: Color.black.opacity(0.25), radius: 10, x: 0, y: 4)
            .animation(.easeOut(duration: 0.18), value: configuration.isPressed)
    }
}

File: android/app/src/main/java/io/talise/app/feature/home/HistoryRow.kt
package io.talise.app.feature.home

import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AccountBalance
import androidx.compose.material.icons.filled.ArrowOutward
import androidx.compose.material.icons.filled.Eco
import androidx.compose.material.icons.filled.SouthWest
import androidx.compose.material.icons.outlined.Circle
import androidx.compose.material.icons.outlined.Eco
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import io.talise.app.R
import io.talise.app.core.model.ActivityEntryDTO
import io.talise.app.ui.components.MicroLabel
import io.talise.app.ui.theme.TaliseColors
import io.talise.app.ui.theme.TaliseType
import androidx.compose.ui.text.font.FontWeight

/**
 * Single history row. Reused by Home (top 4) and HistoryView (full list),
 * a 1:1 port of iOS `HistoryRow`.
 *
 * FLAT row — the enclosing card supplies the surface; the row itself is
 * transparent at rest with a flat solid circular icon chip. On press it picks
 * up a faint directional wash (red for money-out, green for money-in).
 */
@Composable
internal fun HistoryRow(
    entry: ActivityEntryDTO,
    amountsHidden: Boolean,
    onTap: () -> Unit,
) {
    val category = categoryOf(entry)

    // Circular badge fill — a flat colored disc (no glass).
    val badgeBg: Color = when (category) {
        TxCategory.SENT, TxCategory.CASHOUT, TxCategory.TEAM -> HomeSentRed.copy(alpha = 0.16f)
        TxCategory.RECEIVED -> HomeReceivedGreen.copy(alpha = 0.20f)
        TxCategory.INVEST, TxCategory.AUTOSWAP -> TaliseColors.accent.copy(alpha = 0.20f)
        TxCategory.WITHDRAW -> HomeReceivedMint.copy(alpha = 0.42f)
        TxCategory.NEUTRAL -> TaliseColors.surface2
    }
    val badgeFg: Color = when (category) {
        TxCategory.SENT, TxCategory.CASHOUT, TxCategory.TEAM -> HomeSentRedSoft
        TxCategory.RECEIVED -> HomeReceivedMint
        TxCategory.INVEST, TxCategory.AUTOSWAP -> TaliseColors.accent
        TxCategory.WITHDRAW -> HomeWithdrawForest
        TxCategory.NEUTRAL -> TaliseColors.fg
    }
    val icon: ImageVector = when (category) {
        TxCategory.SENT -> Icons.Filled.ArrowOutward
        TxCategory.CASHOUT -> Icons.Filled.AccountBalance
        TxCategory.RECEIVED -> Icons.Filled.SouthWest
        TxCategory.INVEST, TxCategory.AUTOSWAP -> Icons.Filled.Eco
        TxCategory.WITHDRAW -> Icons.Outlined.Eco
        TxCategory.TEAM -> Icons.Filled.AccountBalance // unused — team renders hi_team
        TxCategory.NEUTRAL -> Icons.Outlined.Circle
    }
    // Directional press tint — red for money-out, green for money-in; neutral
    // never tints. Applied only while pressed (iOS HistoryRowButtonStyle).
    val tintColor: Color = when (category) {
        TxCategory.SENT, TxCategory.CASHOUT, TxCategory.TEAM -> HomeSentRed
        TxCategory.RECEIVED, TxCategory.WITHDRAW -> HomeReceivedGreen
        TxCategory.INVEST, TxCategory.AUTOSWAP -> TaliseColors.accent
        TxCategory.NEUTRAL -> TaliseColors.fgMuted
    }
    val tintAlpha = if (category == TxCategory.NEUTRAL) 0f else 0.18f

    val interaction = remember { MutableInteractionSource() }
    val pressed by interaction.collectIsPressedAsState()
    val pressWash by animateColorAsState(
        targetValue = if (pressed) tintColor.copy(alpha = tintAlpha) else Color.Transparent,
        animationSpec = tween(150),
        label = "rowPressWash",
    )

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(14.dp))
            .background(pressWash)
            .clickable(interactionSource = interaction, indication = null) { onTap() }
            .padding(horizontal = 16.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(14.dp),
    ) {
        Box(Modifier.size(36.dp)) {
            Box(
                Modifier.size(36.dp).clip(CircleShape).background(badgeBg),
                contentAlignment = Alignment.Center,
            ) {
                if (category == TxCategory.TEAM) {
                    // Team payouts use the HugeIcons team glyph.
                    Icon(
                        painterResource(R.drawable.hi_team),
                        contentDescription = null,
                        tint = badgeFg,
                        modifier = Modifier.size(18.dp),
                    )
                } else {
                    Icon(icon, contentDescription = null, tint = badgeFg, modifier = Modifier.size(14.dp))
                }
            }
            // Cash-out rows carry the destination country's flag tucked into
            // the bottom-trailing corner of the badge.
            cashoutFlagCode(entry)?.let { code ->
                val flagRes = if (code == "ng") R.drawable.flag_ng else R.drawable.flag_us
                Image(
                    painter = painterResource(flagRes),
                    contentDescription = null,
                    contentScale = ContentScale.Crop,
                    modifier = Modifier
                        .offset(x = 12.dp, y = 12.dp)
                        .size(20.dp)
                        .clip(CircleShape)
                        .border(2.dp, TaliseColors.bg, CircleShape),
                )
            }
        }
        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.dp)) {
            Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(6.dp)) {
                Text(
                    rowTitle(entry, category),
                    style = TaliseType.body(13.sp, FontWeight.Light),
                    letterSpacing = (-0.48).sp,
                    color = TaliseColors.fg,
                )
                // Cash-out rows carry a small disbursement-status pill
                // (Pending / Failed) so the user can tell at a glance
                // whether the naira has landed.
                entry.offramp?.let { off ->
                    offrampStatusPill(off)?.let { (label, color) ->
                        Text(
                            label,
                            style = TaliseType.body(9.sp, FontWeight.SemiBold),
                            letterSpacing = 0.2.sp,
                            color = color,
                            modifier = Modifier
                                .clip(CircleShape)
                                .background(color.copy(alpha = 0.16f))
                                .padding(horizontal = 6.dp, vertical = 2.dp),
                        )
                    }
                }
            }
            MicroLabel(text = rowSubtitle(entry), color = TaliseColors.fgDim)
        }
        // Amount only — the whole row is tappable.
        Text(
            if (amountsHidden) "••••" else rowAmount(entry, category),
            style = TaliseType.body(14.sp, FontWeight.Light),
            letterSpacing = (-0.56).sp,
            color = if (amountsHidden) TaliseColors.fgMuted else rowAmountColor(entry, category, TaliseColors.fg),
        )
    }
}
